pkgs/applications/virtualization/qemu-test/encode-video.c

```diff
@@ -0,0 +1,737 @@
+    }
+
+    if ((context = avcodec_alloc_context3(codec)) == NULL) {
+        fputs("Unable to allocate context for video codec.\n", stderr);
+        goto out_err;
+    }
+
+    if ((bounds = get_bounds(ifile)) == NULL)
+        goto out_err;
+
+    context->width = bounds->width;
+    context->height = bounds->height;
+    start_time = bounds->start_time;
+    frames = bounds->frames;
+
+    free(bounds);
+
+    context->codec_id = fcontext->oformat->video_codec;
+    context->time_base = (AVRational){1, 1000};
+
+    DICT_SET_INT(threads, get_thread_count());
+
+    switch (context->codec_id) {
+        case AV_CODEC_ID_VP8:
+        case AV_CODEC_ID_VP9:
+            /* Encode VP8 and VP9 in constant quality (CQ) mode, so we need to
+             * explicitly set the bit rate to 0.
+             */
+            context->bit_rate = 0;
+            DICT_SET_INT(crf, 30);
+            break;
+        default:
+            context->flags |= CODEC_FLAG_QSCALE;
+            context->global_quality = 1 * FF_QP2LAMBDA;
+            break;
+    }
+
+    if (codec->pix_fmts == NULL) {
+        fprintf(stderr, "Unable to determine pixel format for codec '%s'.\n",
+                avcodec_get_name(fcontext->oformat->video_codec));
+        goto out_err;
+    }
+
+    /* Pick the first supported pixel format of the current codec. */
+    context->pix_fmt = codec->pix_fmts[0];
+
+    if (avcodec_open2(context, codec, &opt) != 0) {
+        fprintf(stderr, "Unable to open context with codec '%s'.\n",
+                avcodec_get_name(fcontext->oformat->video_codec));
+        goto out_err;
+    }
+
+    if (avcodec_parameters_from_context(stream->codecpar, context) < 0) {
+        fputs("Unable to copy stream parameters.\n", stderr);
+        goto out_err;
+    }
+
+    av_dump_format(fcontext, 0, argv[2], 1);
+
+    oframe = alloc_frame(context->pix_fmt, context->width, context->height);
+    if (oframe == NULL)
+        goto out_err;
+
+    if (!maybe_open_avio(fcontext, argv[2]))
+        goto out_err;
+
+    if ((ret = avformat_write_header(fcontext, &opt)) < 0) {
+        fprintf(stderr, "Unable to write stream header to '%s': %s\n",
+                argv[2], av_err2str(ret));
+        goto out_err;
+    }
+
+    if (!encode_frames(ifile, context, fcontext, stream, oframe, start_time,
+                       frames))
+        goto out_err;
+
+    if ((ret = av_write_trailer(fcontext)) < 0) {
+        fprintf(stderr, "Unable to write stream trailer to '%s': %s\n",
+                argv[2], av_err2str(ret));
+        goto out_err;
+    }
+
+    goto out;
+
+out_err:
+    ecode = EXIT_FAILURE;
+
+out:
+    if (oframe != NULL)
+        av_frame_free(&oframe);
+    if (context != NULL)
+        avcodec_free_context(&context);
+    if (fcontext != NULL) {
+        if (fcontext->pb != NULL)
+            maybe_close_avio(fcontext);
+        avformat_free_context(fcontext);
+    }
+    gzclose_r(ifile);
+    av_dict_free(&opt);
+
+    return ecode;
+}
```

Comment:
IMHO the Nixpkgs repo is really not the place to store large C source files.

Reply:
I could also put them into a separate repository if you'd prefer that.